### ISSUE TITLE
failed to push image after GCB build

### ISSUE THREAD

**Issue by cpanato** [kubernetes/k8s.io#1772]
We have a [presubmit prow job](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/release/release-config.yaml#L91-L123) to test if the images are built correctly 

we have an issue that is the job does not complete successfully because it cannot push image we got this error

```
denied: Token exchange failed for project 'k8s-staging-build-image'. Caller does not have permission 'storage.buckets.get'. To configure permissions, follow instructions at: https://cloud.google.com/container-registry/docs/access-control
```
failed job example: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/release/1890/pull-release-image-kube-cross/1359158451615305728

**Comment by spiffxp** [comment]
Ohhhhhh. Ok so here's what's happening.

gcb-builder-releng-test is just used to trigger cloud builds, it is _not_ what those builds run as

AFAIK those always run as a special service account created by GCB in the project that is running those builds

(Dropping k8s-staging since typing on a device) So, I think, the prowjob is running in k8s-infra-prow-build. But its triggering a build in releng-test. That build is running as a different SA in releng-test (https://cloud.google.com/build/docs/cloud-build-service-account#default_permissions_of_service_account). Where it's failing, I think, is trying to push images to some completely different project build-image.

Help me understand what you're trying to do again?

I suspect you do not want presubmits (possibly untrusted code) pushing to build-image, but instead verifying that images can build and push _somewhere_

**Comment by spiffxp** [comment]
Piecemeal configuring certain projects being able to push to certain repos sounds like a recipe for complication. What's the general pattern we're trying to accomplish?

I'm interested in being able to push to buckets/images within a project, whether from prow or from gcb. Moving away from one special SA being able to push to all projects at once.

**Comment by spiffxp** [comment]
tl:dr have you tried pushing to gcr.up/k8s-staging-releng-test instead?

**Comment by justaugustus** [comment]
Another thought...

Ultimately, I'm not sure that we care the entire workflow, just the fact that the image was built successfully.
What about enabling the presubmit/RelEng tools to have a `--local` mode, which would use `cloud-build-local` instead: https://cloud.google.com/build/docs/build-debug-locally

**Comment by justaugustus** [comment]
> tl;dr have you tried pushing to gcr.io/k8s-staging-releng-test instead?

But agreed w/ @spiffxp's assessment, the [most recent (non-go1.16.1) job](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/release/1795/pull-release-image-kube-cross/1367059502620217344) is running in `k8s-staging-releng-test`, but attempting to push to `k8s-staging-build-image`, which is going to fail unless the GCB SA has access to write into GCR for the other project.

If we decide we definitely want to test the push step, then the job should be reconfigured to build and push to the same project.

**Comment by spiffxp** [comment]
You might be able to change your cloudbuild to not push, depending on presence of an env var? And/or if image-builder doesn't support "no-push" that seems like a worthwhile addition

FWIW it may not be a problem for your use case, but bumping into this fun with buildx is why I like the idea of exercising a push

https://github.com/kubernetes/test-infra/issues/20884#issuecomment-783840975

**Comment by cpanato** [comment]
I will check how to set a different registry and the no push option as well

**Comment by cpanato** [comment]
thanks for all the inputs and feedback

**Comment by cpanato** [comment]
after some changes now it works! 🎉 

one GCB build for example: https://console.cloud.google.com/cloud-build/builds;region=global/9e1057f7-eb76-4839-a95f-b34c61815278?project=k8s-staging-releng-test
prow job for that: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/release/1890/pull-release-image-kube-cross/1370028453318889472

we can close this issue

thanks Aaron and Stephen!
/woof
/close
